### PR TITLE
Fix GKD gradient accumulation token loss scaling

### LINKED ISSUE #4719: Loss calculation of `GKDTrainer` may be inaccurate when performing gradient accumulation?
It seems that `GKDTrainer` averages the loss of tokens in a micro batch ahead?

https://github.com/huggingface/trl/blob/8918c9836a3e0b43a6851c08d01b69072f56ca52/trl/experimental/gkd/gkd_trainer.py#L284

### PR DESCRIPTION
**Pull request by taivu1998**
# What does this PR do?

Fixes #4719.

This PR fixes GKD loss scaling when `GKDTrainer` runs with gradient accumulation. The trainer was reducing each microbatch with a local valid-token denominator, so splitting the same logical batch across accumulation steps could change the gradient scale. The fix makes GKD use the accumulation-window token count that `Trainer` computes for custom token-normalized losses.

Root cause:

- `GKDTrainer` did not opt into `Trainer`'s `num_items_in_batch` plumbing for custom loss functions.
- The standard generalized JSD path therefore used local `batchmean` normalization per microbatch instead of the full accumulation-window denominator.
- For on-policy and sequence-level KD, generated labels are created after the dataloader batch is read. Those generated labels must exist before `Trainer` counts valid tokens, and prompt tokens in generated labels must stay masked to `-100`.

Changes:

- Enables custom-loss token counting in `GKDTrainer`.
- Prepares GKD-generated inputs before the accumulation-window denominator is computed.
- Masks prompt labels in generated on-policy outputs.
- Passes `num_items_in_batch` into the generalized JSD loss path and uses it as the global token denominator.
- Preserves the old local reduction when no global denominator is supplied.
- Applies matching scaling for the Liger loss path.
- Adds focused regression tests for accumulated-vs-combined loss scaling, generated-label counting, prompt masking, denominator forwarding, zero-token behavior, and avoiding duplicate generation.

Validation:

- `uv run --extra test pytest tests/experimental/test_gkd_trainer.py -q`
  - `18 passed, 1 skipped`
- `uv run --extra test pytest tests/test_sft_trainer.py::TestDFTLoss -q`
  - `1 passed`
- `uv run --with ruff ruff check trl/experimental/gkd/gkd_trainer.py tests/experimental/test_gkd_trainer.py`
- `uv run --with ruff ruff format --check trl/experimental/gkd/gkd_trainer.py tests/experimental/test_gkd_trainer.py`
- `uv run python -m py_compile trl/experimental/gkd/gkd_trainer.py tests/experimental/test_gkd_trainer.py`
- `git diff --check`

Fixes #4719

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case. See #4719.
- [x] Did you make sure to update the documentation with your changes? No user-facing docs update is needed for this experimental trainer bug fix.
- [x] Did you write any new necessary tests?

## AI writing disclosure

We welcome the use of AI tools to help with contributions. For transparency and to help us improve our review process, please indicate the level of AI involvement in this PR.

- [ ] No AI usage: the PR was written entirely by a human.
- [x] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

Anyone familiar with `GKDTrainer`, `Trainer` loss normalization, or gradient accumulation behavior is welcome to review.

---

> [!NOTE]
> **Medium Risk**
> Touches loss computation and batching/gradient-accumulation accounting in `GKDTrainer`, which can subtly change training dynamics; changes are scoped and covered by targeted regression tests.
> 
> **Overview**
> Fixes GKD loss normalization under gradient accumulation by opting `GKDTrainer` into Trainer’s `num_items_in_batch` plumbing and using that *global* valid-token denominator in `generalized_jsd_loss` (with safe clamping for zero tokens), plus matching rescaling for the Liger fused-loss path.
> 
> Refactors on-policy/seq-KD input generation into `_prepare_gkd_inputs_for_loss`, runs it in `get_batch_samples` so token counting sees the generated/masked labels, and ensures `generate_on_policy_outputs` masks prompt tokens to `-100`; `training_step` now skips duplicate generation via an `_gkd_inputs_prepared` flag. Adds regression tests covering accumulated-vs-combined batch equivalence, zero-token behavior, prompt masking, denominator forwarding, correct token counting after preparation, and avoiding double-preparation.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit b86cb0b27116d483aa1f69eceef10c3d4730852b. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by qgallouedec** [comment]
As mentionned in our contributing guidelines we don't welcome pure AI generated PR. Please stop opening on this repo.
